File: python/Q206_reverse_linked_list.py
#
# @lc app=leetcode id=206 lang=python3
#
# [206] Reverse Linked List
#


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val: int = val
        self.next: ListNode | None = next


# @lc code=start
class Solution:
    def reverseList(self, head: ListNode | None) -> ListNode | None:
        previous: ListNode | None = None
        current: ListNode | None = head

        while current:
            tmp: ListNode | None = current.next
            current.next = previous
            previous = current
            current = tmp

        return previous


# @lc code=end

File: rust/src/leetcode/Q206_reverse_linked_list.rs
/*
 * @lc app=leetcode id=206 lang=rust
 *
 * [206] Reverse Linked List
 */

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @lc code=start
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut previous: Option<Box<ListNode>> = None;
        let mut current: Option<Box<ListNode>> = head;

        while let Some(mut node) = current {
            current = node.next.take();
            node.next = previous;
            previous = Some(node);
        }

        previous
    }
}
// @lc code=end
